src/ray/gcs/gcs_server/gcs_job_manager.cc

```diff
@@ -146,13 +147,66 @@ void GcsJobManager::HandleGetAllJobInfo(rpc::GetAllJobInfoRequest request,
                                         rpc::GetAllJobInfoReply *reply,
                                         rpc::SendReplyCallback send_reply_callback) {
   RAY_LOG(INFO) << "Getting all job info.";
-  auto on_done = [reply, send_reply_callback](
+  auto on_done = [this, reply, send_reply_callback](
                      const absl::flat_hash_map<JobID, JobTableData> &result) {
     for (auto &data : result) {
       reply->add_job_info_list()->CopyFrom(data.second);
     }
-    RAY_LOG(INFO) << "Finished getting all job info.";
-    GCS_RPC_SEND_REPLY(send_reply_callback, reply, Status::OK());
+
+    std::shared_ptr<std::atomic<int>> num_processed_jobs =
+        std::make_shared<std::atomic<int>>(0);
+
+    // For each job with a submission id, we query the internal KV store to get the Ray
+    // Job API JobInfo, and send the gRPC reply in the KV callback. If there are *no* jobs
+    // with a submission id, we must send the gRPC reply in the current function instead.
+    // Use this flag to track it.
+    bool any_jobs_with_submission_id = false;
+
+    for (int i = 0; i < reply->job_info_list_size(); i++) {
+      const auto &metadata = reply->job_info_list(i).config().metadata();
+      if (metadata.find("job_submission_id") == metadata.end()) {
+        num_processed_jobs->fetch_add(1);
+        continue;
+      }
+
+      any_jobs_with_submission_id = true;
+      const auto &job_submission_id = metadata.at("job_submission_id");
+
+      auto kv_get_callback =
+          [reply, send_reply_callback, num_processed_jobs, i, &job_submission_id](
+              std::optional<std::string> job_info_json) {
+            if (job_info_json) {
+              // Parse the JSON into a JobsAPIInfo proto.
+              rpc::JobsAPIInfo jobs_api_info;
+              RAY_CHECK(google::protobuf::util::JsonStringToMessage(*job_info_json,
+                                                                    &jobs_api_info)
+                            .ok());
+
+              // Load info into the reply.
+              reply->mutable_job_info_list(i)->mutable_job_info()->CopyFrom(
+                  jobs_api_info);
+            } else {
+              RAY_LOG(ERROR)
+                  << "Failed to look up Ray Job API JobInfo for job with submission ID "
+                  << job_submission_id;
+            }
+
+            // Send reply if all jobs have been processed.
+            if (num_processed_jobs->fetch_add(1) == reply->job_info_list_size() - 1) {
```

Comment:
nit: I always find post increments a little hard to follow, might be easier to read this way:
```suggestion
            if (++num_processed_jobs == reply->job_info_list_size()) {
```

Reply:
note that `x.fetch_add(1)` is equivalent to `x++`

this might be a little uglier given the shared_ptr though... up to you

Reply:
It does seem a little annoying actually:
`src/ray/gcs/gcs_server/gcs_job_manager.cc:195:17: error: cannot increment value of type 'const std::shared_ptr<std::atomic<int> >'
            if (++num_processed_jobs == reply->job_info_list_size()) {`
            
I think I'll keep it as is